### PR TITLE
Limit num hits when reading Lucene changes

### PR DESCRIPTION
**Pull request by dnhatn**
Today we don't limit the number of hits when reading changes from Lucene
index. If the index and the requesting seq# range both are large, the
searcher may consume a huge amount of memory.

This commit uses a fixed size batch with search_after to avoid the
problem.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/engine/LuceneChangesSnapshot.java [hunk 1]
```diff
@@ -146,28 +148,28 @@ private void rangeCheck(Translog.Operation op) {
...
+        return DocIdSetIterator.NO_MORE_DOCS;
     }
 
-    private TopDocs searchOperations(IndexSearcher searcher) throws IOException {
-        final Query rangeQuery = new BooleanQuery.Builder()
-            .add(new DocValuesFieldExistsQuery(SeqNoFieldMapper.PRIMARY_TERM_NAME), BooleanClause.Occur.FILTER)
-            .add(LongPoint.newRangeQuery(SeqNoFieldMapper.NAME, fromSeqNo, toSeqNo), BooleanClause.Occur.FILTER)
-            .build();
+    private TopDocs searchOperations(ScoreDoc after) throws IOException {
+        final Query rangeQuery = LongPoint.newRangeQuery(SeqNoFieldMapper.NAME, fromSeqNo, toSeqNo);
```
Comment (bleskes) [inline server/src/main/java/org/elasticsearch/index/engine/LuceneChangesSnapshot.java]: why did we drop the  `DocValuesFieldExistsQuery(SeqNoFieldMapper.PRIMARY_TERM_NAME)` part?

**Review by bleskes — COMMENTED** [review]
I left a question.  I was also wondering if we have already a test in place which makes sure that we overflow the `SEARCH_BATCH_SIZE` (as I don't see tests here)?

File: server/src/main/java/org/elasticsearch/index/engine/LuceneChangesSnapshot.java (hunk 1 above)
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/index/engine/LuceneChangesSnapshot.java]: `DocValuesFieldExistsQuery(SeqNoFieldMapper.PRIMARY_TERM_NAME)` was in the temporary fix (https://github.com/elastic/elasticsearch/commit/7004d2c7bd070f49872da6152e05d784fe16b69c). This clause was to use to eliminate nested docs.

File: server/src/main/java/org/elasticsearch/index/engine/LuceneChangesSnapshot.java [hunk 2]
```diff
@@ -55,6 +53,7 @@
  * A {@link Translog.Snapshot} from changes in a Lucene index
  */
 final class LuceneChangesSnapshot implements Translog.Snapshot {
+    static final int SEARCH_BATCH_SIZE = 100;
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/index/engine/LuceneChangesSnapshot.java]: Can this be specified as a parameter via `newLuceneChangesSnapshot(...)` method instead of a constant? Then in ccr we can make this configurable.
 
Also 100 feels on the low side to me. Maybe default to `1024`?

**Review by martijnvg — COMMENTED** [review]
Thanks @dnhatn! I left a comment, also like @bleskes mentioned we need to test fetching of next window in `LuceneChangesSnapshotTests`.

**Comment by dnhatn** [comment]
@bleskes and @martijnvg I've added a test that verifies reading multiple batches. Could you please have another look? Thank you!

File: server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 3]
```diff
@@ -2358,7 +2358,8 @@ long getNumDocUpdates() {
         }
         Searcher searcher = acquireSearcher(source, SearcherScope.INTERNAL);
         try {
-            LuceneChangesSnapshot snapshot = new LuceneChangesSnapshot(searcher, mapperService, minSeqNo, maxSeqNo, requiredFullRange);
+            final int batchSize = preferredSearchBatchSize <= 0 ? Engine.LUCENE_HISTORY_DEFAULT_SEARCH_BATCH_SIZE : preferredSearchBatchSize;
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: I think it is better to define this default just in ccr (`ShardChanges.java`)?
(and let newLuceneChangesSnapshot(...) methods just not allow values lower than 1)

**Review by martijnvg — APPROVED** [review]
Left one comment. Otherwise LGTM

File: server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java (hunk 3 above)
Comment (bleskes) [reply server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: If we do this (and I'm not convinced we should) that we please not use a magic number for the parameter? people should just say what they want or use this constant directly

File: server/src/main/java/org/elasticsearch/index/engine/Engine.java [hunk 4]
```diff
@@ -98,6 +98,8 @@
 
     public static final String SYNC_COMMIT_ID = "sync_id";
     public static final String HISTORY_UUID_KEY = "history_uuid";
+    // The default number of hits that one search should return when reading Lucene history
+    public static final int LUCENE_HISTORY_DEFAULT_SEARCH_BATCH_SIZE = 1024;
```
Comment (bleskes) [inline server/src/main/java/org/elasticsearch/index/engine/Engine.java]: can you move this to LuceneChangesSnapshot ? this will allow using a short name like `DEFAULT_BATCH_SIZE`

File: server/src/test/java/org/elasticsearch/index/engine/LuceneChangesSnapshotTests.java [hunk 5]
```diff
@@ -272,4 +310,8 @@ public void run() {
         }
         return operations;
     }
+
+    private int randomSearchBatchSize() {
```
Comment (bleskes) [inline server/src/test/java/org/elasticsearch/index/engine/LuceneChangesSnapshotTests.java]: we only need a method for this because the default is long. Can we use a short default name or a constant and just inline this.

File: server/src/test/java/org/elasticsearch/index/engine/LuceneChangesSnapshotTests.java [hunk 6]
```diff
@@ -261,6 +271,34 @@ public void run() {
         }
     }
 
+    public void testSearchMultipleBatches() throws Exception {
```
Comment (bleskes) [inline server/src/test/java/org/elasticsearch/index/engine/LuceneChangesSnapshotTests.java]: sine we now randomize the batching in tests, i don't think we need this dedicated test any more?

**Review by bleskes — CHANGES_REQUESTED** [review]
I left some more minor comments and one question I will reach out to discuss.

**Comment by dnhatn** [comment]
I discussed this with @martijnvg and agreed to back out the extra parameter. @bleskes Could you please take a look? Thank you!

**Review by s1monw — APPROVED** [review]
LGTM

**Review by bleskes — APPROVED** [review]
LGTM2

**Comment by dnhatn** [comment]
Thanks everyone!
